numba/tests/test_array_reductions.py

```diff
@@ -435,6 +438,112 @@ def test_nanmedian_basic(self):
         pyfunc = array_nanmedian_global
         self.check_median_basic(pyfunc, self._array_variations)
 
+    def test_fill_diagonal(self):
+        self._fill_diagonal_basic()
+        self._fill_diagonal_exception_cases()
+        self._fill_diagonal_handle_unsafe_type_coercion()
+
+    def _fill_diagonal_basic(self):
+        pyfunc = fill_diagonal_global
+        cfunc = jit(nopython=True)(pyfunc)
+
+        def _shape_variations(n):
+            # square
+            yield (n, n)
+            # tall and thin
+            yield (2 * n, n)
+            # short and fat
+            yield (n, 2 * n)
+            # a bit taller than wide; odd numbers of rows and cols
+            yield ((2 * n + 1), (2 * n - 1))
+            # 4d, all dimensions same
+            yield (n, n, n, n)
+            # weird edge case
+            yield (1, 1, 1)
+
+        def _val_variations():
+            yield 1
+            yield 3.142
+            yield np.nan
+            yield -np.inf
+            yield True
+            yield np.arange(4)
+            yield np.arange(54).reshape(9, 3, 2, 1)
+            yield (4,)
+            yield [8, 9]
+
+        def _multi_dimensional_array_variations(n):
+            for shape in _shape_variations(n):
+                yield np.zeros(shape, dtype=np.float64)
```

Comment:
Might be good to add in something like:
```python
        def _multi_dimensional_array_variations_strided(n):
            for shape in _shape_variations(n):
                tmp = np.zeros(tuple([x * 2 for x in shape]), dtype=np.float64)
                slicer = [slice(0, x * 2, 2) for x in shape]
                yield tmp[slicer]
```
so as to get some slices, same goes for `val`. Might also be worth oscillating layout between 'C' and 'F' in the contig case?

Reply:
Yep, agreed - tests added